overlord/ifacestate/udevmonitor/udevmon.go

```diff
@@ -0,0 +1,195 @@
+		// this cannot happen in real code but may happen in tests
+		return fmt.Errorf("cannot connect: already connected")
+	}
+
+	if err := m.netlinkConn.Connect(netlink.UdevEvent); err != nil {
+		return fmt.Errorf("failed to start uevent monitor: %s", err)
+	}
+
+	// TODO: consider passing a device filter to reduce noise from irrelevant devices.
+	m.monitorStop = m.netlinkConn.Monitor(m.netlinkEvents, m.netlinkErrors, nil)
+
+	return nil
+}
+
+func (m *Monitor) Disconnect() error {
+	close(m.monitorStop)
+	return m.netlinkConn.Close()
+}
+
+// Run enumerates existing USB devices and starts a new goroutine that
+// handles hotplug events (devices added or removed). It returns immediately.
+// The goroutine must be stopped by calling Stop() method.
+func (m *Monitor) Run() error {
+	m.tomb.Go(func() error {
+		existingDevices := make(chan *hotplug.HotplugDeviceInfo)
+		udevadmErrors := make(chan error)
+
+		// Gather devices from udevadm info output (enumeration on startup).
+		if err := hotplug.EnumerateExistingDevices(existingDevices, udevadmErrors); err != nil {
+			return fmt.Errorf("cannot enumerate existing devices: %s", err)
+		}
+		var finished bool
+		for !finished {
+			select {
+			case dev, ok := <-existingDevices:
+				if dev != nil {
+					if _, seen := m.devices[dev.DevicePath()]; seen {
+						break
+					}
+					m.devices[dev.DevicePath()] = true
+					if ok && m.deviceAdded != nil {
+						m.deviceAdded(dev)
+					}
+				}
+				if !ok {
+					finished = true
+				}
+			case err, ok := <-udevadmErrors:
+				if ok {
+					logger.Noticef("udevadm error: %q\n", err)
+				} else {
+					udevadmErrors = nil
+				}
+			case <-m.tomb.Dying():
+				return m.Disconnect()
+			}
+		}
+
+		// Process hotplug events reported by udev monitor.
+		for {
+			select {
+			case err := <-m.netlinkErrors:
+				logger.Noticef("netlink error: %q\n", err)
+			case ev := <-m.netlinkEvents:
+				m.udevEvent(&ev)
+			case <-m.tomb.Dying():
+				return m.Disconnect()
+			}
+		}
+	})
+
+	return nil
+}
+
+func (m *Monitor) Stop() error {
+	m.tomb.Kill(nil)
+	err := m.tomb.Wait()
+	m.netlinkConn = nil
+	return err
+}
+
+func (m *Monitor) udevEvent(ev *netlink.UEvent) {
+	switch ev.Action {
+	case netlink.ADD:
+		m.addDevice(ev.KObj, ev.Env)
+	case netlink.REMOVE:
+		m.removeDevice(ev.KObj, ev.Env)
+	default:
+	}
+}
+
+func (m *Monitor) addDevice(kobj string, env map[string]string) {
+	dev, err := hotplug.NewHotplugDeviceInfo(env)
+	if err != nil {
+		return
+	}
+	if _, seen := m.devices[dev.DevicePath()]; seen {
+		return
+	}
+	m.devices[dev.DevicePath()] = true
+	if m.deviceAdded != nil {
+		m.deviceAdded(dev)
+	}
+}
+
+func (m *Monitor) removeDevice(kobj string, env map[string]string) {
+	dev, err := hotplug.NewHotplugDeviceInfo(env)
+	if err != nil {
+		return
+	}
+	if _, seen := m.devices[dev.DevicePath()]; !seen {
+		logger.Noticef("received device remove event, but device %q was not seen before", dev.DevicePath())
```

Comment:
Perhaps some more context:

"udev monitor observed remove event for unknown device %q"

Reply:
Done.